engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractors.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.engine.cascading;
+
+import java.lang.reflect.TypeVariable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+import org.hibernate.validator.internal.util.TypeHelper;
+import org.hibernate.validator.internal.util.TypeVariableBindings;
+import org.hibernate.validator.spi.cascading.ValueExtractor;
+
+/**
+ * @author Gunnar Morling
+ */
+public class ValueExtractors {
+
+	private final List<ValueExtractorDescriptor> valueExtractors;
+
+	public ValueExtractors(Iterable<ValueExtractor<?>> externalExtractors) {
+		List<ValueExtractorDescriptor> tmpValueExtractors = new ArrayList<>();
+
+		for ( ValueExtractor<?> valueExtractor : externalExtractors ) {
+			tmpValueExtractors.add( new ValueExtractorDescriptor( valueExtractor ) );
+		}
+
+		tmpValueExtractors.add( new ValueExtractorDescriptor( LegacyListValueExtractor.INSTANCE ) );
+		tmpValueExtractors.add( new ValueExtractorDescriptor( ListValueExtractor.INSTANCE ) );
+
+		tmpValueExtractors.add( new ValueExtractorDescriptor( ArrayValueExtractor.INSTANCE ) );
+
+		tmpValueExtractors.add( new ValueExtractorDescriptor( LegacyMapValueExtractor.INSTANCE ) );
+		tmpValueExtractors.add( new ValueExtractorDescriptor( MapValueExtractor.INSTANCE ) );
+
+		tmpValueExtractors.add( new ValueExtractorDescriptor( LegacyIterableValueExtractor.INSTANCE ) );
+		tmpValueExtractors.add( new ValueExtractorDescriptor( IterableValueExtractor.INSTANCE ) );
+
+		tmpValueExtractors.add( new ValueExtractorDescriptor( ObjectValueExtractor.INSTANCE ) );
+
+		valueExtractors = Collections.unmodifiableList( tmpValueExtractors );
+	}
+
+	public List<ValueExtractor<?>> getCascadedValueExtractors(Class<?> valueType, Class<?> declarationType, List<TypeVariable<?>> cascadingTypeParameters) {
```

Comment:
That thing is nightmare-ish. It works for the general case now, but eventually, the logic would have to be refined. Specifically, it works due to the order of registration of built-in extractors (more specific ones before more general ones). It didn't seem needed to implement it 100% for now.